src/sspi/negotiate.rs

```diff
@@ -69,88 +91,113 @@ pub struct Negotiate {
-        } else {
-            NegotiatedProtocol::Ntlm(Ntlm::new())
-        };
-
-        if let Some(filtered_protocol) = Self::filter_protocol(&protocol, &config.package_list) {
+        let mut protocol = config.protocol_config.new_client()?;
+        if let Some(filtered_protocol) = Self::filter_protocol(&protocol, &config.package_list)? {
             protocol = filtered_protocol;
         }
 
         Ok(Negotiate {
-            protocol: protocol,
-            package_list: config.package_list.clone(),
+            protocol,
+            package_list: config.package_list,
             auth_identity: None,
         })
     }
 
-    #[cfg(feature = "network_client")]
+    // negotiates the authorization protocol based on the username and the domain
+    // Decision rules:
+    // 1) if `self.protocol` is not NTLM then we've already negotiated a suitable protocol. Nothing to do.
+    // 2) if the provided domain is Azure AD domain then it'll use Pku2u
+    // 3) if the provided username is FQDN and we can resolve KDC then it'll use Kerberos
+    // 4) if SSPI_KDC_URL_ENV is set then it'll also use Kerberos
+    // 5) in any other cases, it'll use NTLM
     fn negotiate_protocol(&mut self, username: &str, domain: &str) -> Result<()> {
         if let NegotiatedProtocol::Ntlm(_) = &self.protocol {
-            let realm = get_client_principal_realm(username, domain);
-            if let Some(kdc_url) = detect_kdc_url(&realm) {
-                self.protocol = NegotiatedProtocol::Kerberos(Kerberos::new_client_from_config(KerberosConfig {
-                    url: Some(kdc_url),
-                    network_client: Box::new(ReqwestNetworkClient::new()),
-                })?)
+            #[cfg(target_os = "windows")]
+            if is_azure_ad_domain(domain) {
+                use super::pku2u::Pku2uConfig;
+
+                self.protocol =
+                    NegotiatedProtocol::Pku2u(Pku2u::new_client_from_config(Pku2uConfig::default_client_config()?)?);
+            }
+
+            #[cfg(feature = "network_client")]
+            if let Some(host) = detect_kdc_url(&get_client_principal_realm(username, domain)) {
+                self.protocol =
+                    NegotiatedProtocol::Kerberos(Kerberos::new_client_from_config(crate::KerberosConfig {
+                        url: Some(host),
+                        network_client: Box::new(ReqwestNetworkClient::new()),
+                    })?);
             }
         }
 
-        if let Some(filtered_protocol) = Self::filter_protocol(&self.protocol, &self.package_list) {
+        if let Some(filtered_protocol) = Self::filter_protocol(&self.protocol, &self.package_list)? {
             self.protocol = filtered_protocol;
         }
 
         Ok(())
     }
 
-    fn get_package_list_config(package_list: &Option<String>) -> (bool, bool) {
+    fn get_package_list_config(package_list: &Option<String>) -> (bool, bool, bool) {
```

Comment:
Nitpick. It’s hard for me to understand the return value of this function without looking inside. I think it would help to return a struct with named fields instead. What do you think?

Something in the lines of:

```rust
struct PackageListConfig {
    ntlm_package: bool,
    kerberos_package: bool,
    pku2u_package: bool,
}
```

It’s not a blocker since this is internal API.

Reply:
yes, a struct with protocol names is much better than a tuple with three bools.